include/ghex/structured/rma_range_generator.hpp

```diff
@@ -0,0 +1,190 @@
+            m_local_guard.end_target_epoch();
+        }
+    };
+
+    /** @brief This class represents the source range of a halo exchange operation. It is
+     * referencing a local source field (starting point of the put) to which it has direct memory
+     * access. During construction and recv() member function, it receives a target range from its
+     * remote partner and deserializes it into a generic range type, which exposes synchronization
+     * functions for RMA. The actual type can be recovered from generic range type by type injection
+     * through the range_factory.
+     * @tparam RangeFactory the factory type which knows about all possible range types
+     * @tparam Communicator the communicator type */
+    template<typename RangeFactory, typename Communicator>
+    struct source_range
+    {
+        using field_type = Field;
+        using info = rma::info;
+        using rank_type = typename Communicator::rank_type;
+        using tag_type = typename Communicator::tag_type;
+
+        Communicator m_comm;
+        range_type m_local_range;
+        typename RangeFactory::range_type m_remote_range;
+        rank_type m_src;
+        tag_type m_tag;
+        typename Communicator::template future<void> m_request;
+        std::vector<unsigned char> m_archive;
+
+        template<typename IterationSpace>
+        source_range(const Communicator& comm, const Field& f,
+            const IterationSpace& is, rank_type src, tag_type tag)
+        : m_comm{comm}
+        , m_local_range{f, is.local().first(), is.local().last()-is.local().first()+1}
+        , m_src{src}
+        , m_tag{tag}
+        {
+            m_archive.resize(RangeFactory::serial_size);
+            m_request = m_comm.recv(m_archive, m_src, m_tag);
+        }
+
+        void recv()
+        {
+            m_request.wait();
+            // creates a traget range
+            m_remote_range = RangeFactory::deserialize(m_archive.data());
+            RangeFactory::call_back_with_type(m_remote_range, [this] (auto& r)
+            {
+                init(r, m_remote_range);
+            });
+        }
+        
+        void start_source_epoch()
+        {
+            m_remote_range.start_source_epoch();
+        }
+
+        void end_source_epoch()
+        {
+            m_remote_range.end_source_epoch();
+        }
+
+        void put()
+        {
+            RangeFactory::call_back_with_type(m_remote_range, [this] (auto& r)
+            {
+                put(r);
+            });
+        }
+
+        template<typename TargetRange>
+        void put(TargetRange& tr)
+        {
+            ::gridtools::ghex::structured::put(m_local_range, tr);
+        }
+
+    private:
+        template<typename TargetRange>
+        void init(TargetRange& tr, rma::range& r)
+        {
+            using T = typename TargetRange::value_type;
+            tr.m_field.set_data((T*)r.get_ptr());
+        }
+    };
+};
+
+} // namespace structured
+
+namespace rma {
```

Comment:
I need a clarification here: the traits here are specialized for the structured case. Is this the only case handled right now?

Reply:
yes, this is the only case handled right now.